### PR TITLE
Speedup analysis and testing by parallelizing functions wherever possible. Also implement fast hyperparameter tuning with hyperopt.

### PR DESCRIPTION
**Pull request by harisriaz17**
This pull request has two components:

1. The first one involves parallelizing the following functions:

- Inside mlperformance.py:

- cross_validation_mse, cross_validation_mse_continuous, confusion and precision_recall.

- Inside gridsearch.py:

- do_gridsearch.

- As a result, all analysis methods for LcpK0s_multiplicity (excluding grid search and train) run in under 10 minutes.

- Grid search, even when using all cores takes up to 5 hours when running on a sample of 500k events.

- Additionally, this PR fixes the confusion matrix function where initially the labels for signal and event were swapped. This has now been resolved. The confusion() function now also plots matrices for both train and test sets.

2. The second component aims to provide a generic method for fast (bayesian) hyperparameter tuning for any model over any search space.

- This is accomplished with the use of the external hyperopt (https://github.com/hyperopt/hyperopt) library. As a result, for a sample training set of 500k events, I was able to tune all the parameters of an XGBoost model in ~ 47 minutes.

- Previously the same process would take more than 24 hours with traditional random or grid searches.

- Additionally, the new set of parameters also improved the model's predictions on both train and test sets. Using the F1 score as an evaluation metric (this shall be our preferred metric of choice from now on), I was able to achieve a performance uplift of 7.8% in predicting signals (minority class), whilst also reducing false negatives and false positives.

- Further commits to follow...

### CODE REVIEW

**Comment by ginnocen** [comment]
Hi Haris, I am very much in favor of merging this PR, but I think we would have to make an effort to fix the conflicts and discuss the results in a shared meeting. As you can see it involves several changes in the code, so we have to make sure this doesn't break anything in the workflow. I think you should join one of our regular group meeting and describe the results and the changes. We could schedule it for 1 week from now or 2 weeks from now if you agree. Please be aware that before merging we should check the analysis until its very end, and compare to the performances of the current optimization. @benedikt-voelkel @DelloStritto @vkucera

**Comment by harisriaz17** [comment]
@ginnocen Ok

**Comment by ginnocen** [comment]
hi @harisriaz17 have you managed to polish this PR and solve the conflicts? So that we can finally merge it and start using it. Thanks!

**Comment by ginnocen** [comment]
Superseded by #629
